### PR TITLE
feat: refactor string package reverse

### PR DESCRIPTION
**Pull request by steff456**
Part of https://github.com/stdlib-js/stdlib/issues/1062

## Description

> What is the purpose of this pull request?

This pull request:

- [x] Refactors the string package `@stdlib/string/reverse`
- [x] Adds `@stdlib/string/base/reverse`
- [x] Adds `@stdlib/string/base/reverse-code-points`
- [x] Adds `@stdlib/string/base/reverse-grapheme-clusters`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   Is part of https://github.com/stdlib-js/stdlib/issues/1062

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

**Comment by kgryte** [comment]
@steff456 Looks like the tests don't cover all branches for reversing code points. Would you mind adding a few tests to address?

**Comment by kgryte** [comment]
You can generate a test coverage report locally using `make test-cov`.

**Comment by Planeshifter** [comment]
@steff456 @kgryte 

Ran `make test-cov TESTS_FILTER='.*/string/base/reverse-code-point/.*'` locally and am getting the same coverage as in the table.

One can see the uncovered lines online as well: https://coverage.stdlib.io/string/base/reverse-code-point/

**Comment by steff456** [comment]
Thanks @Planeshifter and @kgryte, I'll check the tests :)

**Comment by steff456** [comment]
I'm not sure why the coverage tests failed with the last commit, but when I try it locally it is working 🙃

**Comment by kgryte** [comment]
@steff456 I believe the test coverage failure is due to a known issue. @Planeshifter can you comment?

**Comment by Planeshifter** [comment]
@steff456 Will look into it, sorry for that; seems like the job was not able to push the coverage report to the test coverage repo. The table now shows that 100% coverage is achieved, though, so that's awesome!

**Comment by steff456** [comment]
Perf! Thanks to both 👏🏻

**Comment by kgryte** [comment]
@steff456 Would you mind renaming the packages from 

```
reverse-code-point
reverse-grapheme-cluster
```

to 

```
reverse-code-points
reverse-grapheme-clusters
```

The use of singular is a bit confusing in this context, as, e.g., `reverse-code-point`, makes it seem as those the function only reverses a single code point, rather than a sequence of code points.

File: lib/node_modules/@stdlib/string/base/reverse-code-points/README.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+// returns 'noitcele laitnediserp'
+
+str = reverseCodePoints( 'JavaScript' );
+// returns 'tpircSavaJ'
+
+str = reverseCodePoints( 'The Last of the Mohicans' );
+// returns 'snacihoM eht fo tsaL ehT'
+
+str = reverseCodePoints( 'अनुच्छेद' );
+// returns 'देछ्चुनअ'
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/base/reverse-code-points/README.md:71]: This looks off to me. Is the reason why some visual characters are different due to some (all?) of them being grapheme clusters?

File: lib/node_modules/@stdlib/string/base/reverse-code-points/examples/index.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+// => 'noitcele laitnediserp'
+
+console.log( reverse( 'JavaScript' ) );
+// => 'tpircSavaJ'
+
+console.log( reverse( 'The Last of the Mohicans' ) );
+// => 'snacihoM eht fo tsaL ehT'
+
+console.log( reverse( 'अनुच्छेद' ) );
+// => 'देछ्चुनअ'
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/base/reverse-code-points/examples/index.js:33]: Same comment as README.

File: lib/node_modules/@stdlib/string/base/reverse-code-points/lib/main.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+			if ( RE_UTF16_LOW_SURROGATE.test( ch2 ) ) {
+				// We found a surrogate pair:
+				out = ch1 + ch2 + out;
+				i += 1; // bump the index to process the next code unit
+			}
+			else {
+				out = ch1 + out;
+			}
+		}
+		else {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/base/reverse-code-points/lib/main.js]:
```suggestion
			} else {
				out = ch1 + out;
			}
		} else {
```

File: lib/node_modules/@stdlib/string/base/reverse-code-points/test/test.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+	t.strictEqual( typeof reverse, 'function', 'main export is a function' );
+	t.end();
+});
+
+tape( 'the function returns an empty string if provided an empty string', function test( t ) {
+	t.strictEqual( reverse( '' ), '', 'returns expected value' );
+	t.end();
+});
+
+tape( 'the function reverses the Unicode code point of a provided string (ascii)', function test( t ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/base/reverse-code-points/test/test.js]:
```suggestion
tape( 'the function reverses the Unicode code points of a provided string (ascii)', function test( t ) {
```

File: lib/node_modules/@stdlib/string/base/reverse-code-points/test/test.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+	out = reverse( '!!!*' );
+	t.strictEqual( out, '*!!!', 'returns expected value' );
+
+	out = reverse( 'Hello World' );
+	t.strictEqual( out, 'dlroW olleH', 'returns expected value' );
+
+	t.end();
+});
+
+tape( 'the function reverses the Unicode code point of a provided string (Unicode)', function test( t ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/base/reverse-code-points/test/test.js]:
```suggestion
tape( 'the function reverses the Unicode code points of a provided string (Unicode)', function test( t ) {
```

File: lib/node_modules/@stdlib/string/base/reverse-code-points/test/test.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+	var out;
+
+	out = reverse( 'अनुच्छेद' );
+	t.strictEqual( out, 'देछ्चुनअ', 'returns expected value' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/base/reverse-code-points/test/test.js:59]: Same comment as README.

File: lib/node_modules/@stdlib/string/base/reverse-code-points/test/test.js [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+	out = reverse( '六书/六書' );
+	t.strictEqual( out, '書六/书六', 'returns expected value' );
+
+	t.end();
+});
+
+tape( 'the function reverses the Unicode code point of a provided string (surrogate pair)', function test( t ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/base/reverse-code-points/test/test.js]:
```suggestion
tape( 'the function reverses the Unicode code points of a provided string (surrogate pair)', function test( t ) {
```

File: lib/node_modules/@stdlib/string/base/reverse-code-points/test/test.js [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+	var out;
+
+	out = reverse( 'foo\uD834\uDF06bar' );
+	t.strictEqual( out, 'rab\uD834\uDF06oof', 'returns expected value' );
+
+	out = reverse( 'foo\uD834bar\uDF06baz' );
+	t.strictEqual( out, 'zab\uDF06rab\uD834oof', 'returns expected value' );
+
+	out = reverse( 'foo\uD834bar\uDF06\uD834' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/base/reverse-code-points/test/test.js:76]: To be clear, this is testing for a trailing unpaired surrogate in addition to a surrogate pair, correct?

File: lib/node_modules/@stdlib/string/base/reverse-grapheme-clusters/README.md [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+# reverseGraphemeClusters
+
+> Reverse the grapheme clusters (i.e., user-perceived characters) of a string.
+
+<section class="usage">
+
+## Usage
+
+<!-- eslint-disable id-length -->
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/base/reverse-grapheme-clusters/README.md]: Does `reverseGraphemeClusters` exceed the maximum id length? If I count correctly, `reverseGraphemeClusters` is only 23 characters which is less than the maximum id length.

If I am correct, we can remove the disabled lint rule comments.

File: lib/node_modules/@stdlib/string/base/reverse-grapheme-clusters/lib/main.js [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+	var brk;
+	var i;
+
+	out = '';
+	idx = str.length - 1;
+	while ( idx >= 0 ) {
+		brk = prevGraphemeClusterBreak( str, idx );
+		for ( i = brk + 1; i <= idx; i++ ) {
+			out += str.charAt( i );
+		}
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/base/reverse-grapheme-clusters/lib/main.js]: Would it make more sense to use `nextGraphemeClusterBreak` and use the prepending logic you used in `reverse-code-points`? My sense is that we'd get better performance due to how `prevGraphemeClusterBreak` works.

File: lib/node_modules/@stdlib/string/base/reverse-grapheme-clusters/package.json [hunk 11]
```diff
@@ -0,0 +1,68 @@
+{
+  "name": "@stdlib/string/base/first-grapheme-clusters",
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/base/reverse-grapheme-clusters/package.json]:
```suggestion
  "name": "@stdlib/string/base/reverse-grapheme-clusters",
```

File: lib/node_modules/@stdlib/string/base/reverse-grapheme-clusters/package.json [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+  "version": "0.0.0",
+  "description": "Reverse the grapheme cluster (i.e., user-perceived character) of a string.",
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/base/reverse-grapheme-clusters/package.json]:
```suggestion
  "description": "Reverse the grapheme clusters (i.e., user-perceived characters) of a string.",
```

File: lib/node_modules/@stdlib/string/base/reverse-grapheme-clusters/test/test.js [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+	t.strictEqual( out, 'dlroW olleH', 'returns expected value' );
+
+	t.end();
+});
+
+tape( 'the function reverses the grapheme clusters of a provided string (Unicode)', function test( t ) {
+	var out;
+
+	out = reverse( 'अनुच्छेद', 1 );
+	t.strictEqual( out, 'दछेच्नुअ', 'returns expected value' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/base/reverse-grapheme-clusters/test/test.js:59]: I think this answers my question above for `reverse-code-points`. This string is comprised of grapheme clusters, not just surrogate pairs.

File: lib/node_modules/@stdlib/string/base/reverse/lib/main.js [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+* @example
+* var out = reverse( 'Hidden Treasures' );
+* // returns 'serusaerT neddiH'
+*/
+function reverse( str ) {
+	var out;
+	var i;
+
+	out = '';
+	for ( i = str.length - 1; i >= 0; i--) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/base/reverse/lib/main.js]:
```suggestion
	for ( i = str.length - 1; i >= 0; i-- ) {
```

File: lib/node_modules/@stdlib/string/base/reverse/lib/main.js [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+		out += str[i];
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/base/reverse/lib/main.js]:
```suggestion
		out += str[ i ];
```

File: lib/node_modules/@stdlib/string/reverse/README.md [hunk 16]
```diff
@@ -45,12 +45,28 @@ out = reverseString( 'Lorem ipsum 𝌆 dolor sit ameͨ͆t.' );
 // returns '.teͨ͆ma tis rolod 𝌆 muspi meroL'
 ```
 
+The function supports the following options:
+
+-   **mode**: type of characters to return. Must be one of the following:
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/reverse/README.md]:
```suggestion
-   **mode**: type of characters to reverse. Must be one of the following:
```

File: lib/node_modules/@stdlib/string/reverse/docs/repl.txt [hunk 17]
```diff
@@ -1,12 +1,31 @@
 
-{{alias}}( str )
+{{alias}}( str[, options] )
     Reverses a `string`.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/reverse/docs/repl.txt]:
```suggestion
    Reverses a string.
```

File: lib/node_modules/@stdlib/string/reverse/docs/repl.txt [hunk 18]
```diff
@@ -1,12 +1,31 @@
...
     Parameters
     ----------
     str: string
         Input string.
 
+    options: Object (optional)
+        Options.
+
+    options.mode: string (optional)
+        Type of characters to return. The following modes are supported:
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/reverse/docs/repl.txt]:
```suggestion
        Type of characters to reverse. The following modes are supported:
```

File: lib/node_modules/@stdlib/string/reverse/docs/types/index.d.ts [hunk 19]
```diff
@@ -18,29 +18,46 @@
...
 // TypeScript Version: 2.0
 
+// tslint:disable:unified-signatures
+
+/**
+* Interface describing function options.
+*/
+interface Options {
+	/**
+	* Specifies the type of characters to return (default: 'grapheme').
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/reverse/docs/types/index.d.ts]:
```suggestion
	* Specifies the type of characters to reverse (default: 'grapheme').
```

File: lib/node_modules/@stdlib/string/reverse/docs/types/index.d.ts [hunk 20]
```diff
@@ -18,29 +18,46 @@
...
+	*
+	* ## Notes
+	*
+	* -   The following option values are supported:
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/reverse/docs/types/index.d.ts]:
```suggestion
	* -   The following modes are supported:
```

File: lib/node_modules/@stdlib/string/reverse/docs/types/test.ts [hunk 21]
```diff
@@ -36,6 +36,33 @@ import reverseString = require( './index' );
...
+	reverseString( true, {} ); // $ExpectError
+	reverseString( false, {} ); // $ExpectError
+	reverseString( null, {} ); // $ExpectError
+	reverseString( undefined, {} ); // $ExpectError
+	reverseString( 5, {} ); // $ExpectError
+	reverseString( [], {} ); // $ExpectError
+	reverseString( {}, {} ); // $ExpectError
+	reverseString( ( x: number ): number => x, {} ); // $ExpectError
+}
+
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/reverse/docs/types/test.ts:49]: We should also add tests here when not provided an object for the second argument.

File: lib/node_modules/@stdlib/string/reverse/docs/usage.txt [hunk 22]
```diff
@@ -6,3 +6,4 @@ Options:
   -h,    --help                Print this message.
   -V,    --version             Print the package version.
          --split sep           Delimiter for stdin data. Default: '/\\r?\\n/'.
+         --mode mode           Type of character to return. Default: 'grapheme'.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/reverse/docs/usage.txt]:
```suggestion
         --mode mode           Type of character to reverse. Default: 'grapheme'.
```

File: lib/node_modules/@stdlib/string/reverse/lib/main.js [hunk 23]
```diff
@@ -20,18 +20,37 @@
...
+
+
 // MAIN //
 
 /**
 * Reverses a string.
 *
 * @param {string} str - input string
+* @param {Options} [options] - options
 * @throws {TypeError} must provide a string primitive
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/reverse/lib/main.js:52]: We're missing the documentation of the `mode` option following L50.

File: lib/node_modules/@stdlib/string/reverse/test/test.cli.js [hunk 24]
```diff
@@ -162,6 +162,46 @@ tape( 'the command-line interface reverses a string argument', opts, function te
 	}
 });
 
+tape( 'the command-line interface supports specifying the type of characters to return', opts, function test( t ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/reverse/test/test.cli.js]:
```suggestion
tape( 'the command-line interface supports specifying the type of characters to reverse', opts, function test( t ) {
```

File: lib/node_modules/@stdlib/string/reverse/test/test.js [hunk 25]
```diff
@@ -116,3 +200,82 @@ tape( 'the function reverses a string', function test( t ) {
...
+	t.strictEqual( out, 'दछेच्नुअ', 'returns expected value' );
+
+	out = reverseString( '六书/六書', opts );
+	t.strictEqual( out, '書六/书六', 'returns expected value' );
+
+	out = reverseString( '🌷', opts );
+	t.strictEqual( out, '🌷', 'returns expected value' );
+
+	out = reverseString( '👉🏿', opts );
+	t.strictEqual( out, '🏿👉', 'returns expected value' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/reverse/test/test.js]: This looks off to me. I'd expect that, similar to L225, the input string equals the output string. Maybe this stems from using `prevGraphemeClusterBreak`?

File: lib/node_modules/@stdlib/string/base/reverse-code-points/examples/index.js (hunk 2 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/string/base/reverse-code-points/examples/index.js:33]: Answered below when reviewing `reverse-grapheme-clusters`.

File: lib/node_modules/@stdlib/string/base/reverse-code-points/README.md (hunk 1 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/string/base/reverse-code-points/README.md:71]: Answered below when reviewing `reverse-grapheme-clusters`.

File: lib/node_modules/@stdlib/string/base/reverse-code-points/test/test.js (hunk 6 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/string/base/reverse-code-points/test/test.js:59]: Answered below when reviewing `reverse-grapheme-clusters`.

File: lib/node_modules/@stdlib/string/base/reverse-grapheme-clusters/README.md [hunk 26]
```diff
@@ -0,0 +1,98 @@
...
+
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/base/reverse-grapheme-clusters/README.md]:
```suggestion
```

File: lib/node_modules/@stdlib/string/base/reverse-grapheme-clusters/README.md [hunk 27]
```diff
@@ -0,0 +1,98 @@
...
+```javascript
+var reverseGraphemeClusters = require( '@stdlib/string/base/reverse-grapheme-clusters' );
+```
+
+#### reverseGraphemeClusters( str )
+
+Reverses the grapheme clusters (i.e., user-perceived characters) of a string.
+
+<!-- eslint-disable id-length -->
+
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/base/reverse-grapheme-clusters/README.md]:
```suggestion
```

File: lib/node_modules/@stdlib/string/base/reverse-grapheme-clusters/README.md [hunk 28]
```diff
@@ -0,0 +1,98 @@
...
+<!-- /.usage -->
+
+<section class="examples">
+
+## Examples
+
+<!-- eslint no-undef: "error" -->
+
+<!-- eslint-disable id-length -->
+
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/base/reverse-grapheme-clusters/README.md]:
```suggestion
```

File: lib/node_modules/@stdlib/string/reverse/docs/types/index.d.ts [hunk 29]
```diff
@@ -18,29 +18,46 @@
 
 // TypeScript Version: 2.0
 
+// tslint:disable:unified-signatures
+
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/reverse/docs/types/index.d.ts]:
```suggestion
```

This is not applicable.

**Review by kgryte — CHANGES_REQUESTED** [review]
@steff456 Thanks for working on this. Left a few comments. The handling of grapheme clusters looks off to me, based on the expected test values. Additionally, we need to add a couple more TypeScript tests and fix some missing documentation.

File: lib/node_modules/@stdlib/string/base/reverse-code-points/test/test.js (hunk 8 above)
Comment (steff456) [reply lib/node_modules/@stdlib/string/base/reverse-code-points/test/test.js:76]: Yes!

File: lib/node_modules/@stdlib/string/base/reverse-grapheme-clusters/lib/main.js [hunk 30]
```diff
@@ -0,0 +1,81 @@
...
+	while ( idx < str.length ) {
+		brk = nextGraphemeClusterBreak( str, idx );
+		if ( brk === -1 ) {
+			brk = str.length;
+		}
+		cluster = '';
+		for ( i = idx; i < brk; i++ ) {
+			cluster += str.charAt( i );
+		}
+		out = cluster + out;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/base/reverse-grapheme-clusters/lib/main.js:72]: For future reference, this is not the most memory efficient approach (i.e., prepending strings), but we don't have much choice until the bug in `prevGraphemeClusterBreak` is resolved (ref: https://github.com/stdlib-js/stdlib/issues/1092).

File: lib/node_modules/@stdlib/string/reverse/docs/types/test.ts [hunk 31]
```diff
@@ -36,6 +36,43 @@ import reverseString = require( './index' );
...
+	reverseString( 'abc', { 'mode': [] } ); // $ExpectError
+	reverseString( 'abc', { 'mode': ( x: number ): number => x } ); // $ExpectError
+
+	reverseString( 'abc', 1, { 'mode': true } ); // $ExpectError
+	reverseString( 'abc', 1, { 'mode': false } ); // $ExpectError
+	reverseString( 'abc', 1, { 'mode': null } ); // $ExpectError
+	reverseString( 'abc', 1, { 'mode': '' } ); // $ExpectError
+	reverseString( 'abc', 1, { 'mode': [] } ); // $ExpectError
+	reverseString( 'abc', 1, { 'mode': {} } ); // $ExpectError
+	reverseString( 'abc', 1, { 'mode': ( x: number ): number => x } ); // $ExpectError
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/reverse/docs/types/test.ts]:
```suggestion
```

These tests do not apply. `reverse` is a variadic function supporting at most 2 arguments.

File: lib/node_modules/@stdlib/string/reverse/docs/types/test.ts [hunk 32]
```diff
@@ -36,6 +36,43 @@ import reverseString = require( './index' );
...
+	reverseString( undefined, {} ); // $ExpectError
+	reverseString( 5, {} ); // $ExpectError
+	reverseString( [], {} ); // $ExpectError
+	reverseString( {}, {} ); // $ExpectError
+	reverseString( ( x: number ): number => x, {} ); // $ExpectError
+}
+
+// The compiler throws an error if the function is provided a value other than an object as second argument...
+{
+	reverseString( 'abc', true ); // $ExpectError
```
Comment (kgryte) [inline lib/node_modules/@stdlib/string/reverse/docs/types/test.ts:54]:
```suggestion
	reverseString( 'abc', '1' ); // $ExpectError
	reverseString( 'abc', 1 ); // $ExpectError
	reverseString( 'abc', true ); // $ExpectError
```

**Review by kgryte — APPROVED** [review]
LGTM. Thanks, @steff456!

**Comment by kgryte** [comment]
The failing CI jobs do not appear related to this PR. As such, will go ahead and merge...

**Pull request merged**